Ledger Live App

michalsmiarowski opened the pull request:
<h1 align="center" fontSize="30">Ledger Live App</h1>

Closes: #649 
~Blocked by: #654~

This PR allows to run our dApp as Live App withing Ledger Live. The Live Apps are displayed in the Discover section of Ledger Live on Desktop (Windows, Mac, Linux) and mobile (Android and iOS).

The main purpose of it would be to complete the whole Mint & Unmint flow, without the need to leave the Ledger Live application and do a bitcoin transaction to generated deposit address. All transactions are done within the application.

# Overall Description

When running as Ledger Live App, our Token Dashboard is embedded into it and displayet differently than in the website. We are checking that with our `?embed=true` query parameter, that I've put in the manifest file. Only tbtc section is needed for this, so that's why onli this section is displayed and the rest are hidden.

The user can connect his ethereum account from Ledger to communicate with eth contracts. He can also choose which of his bitcoin addresses he wants to use to send the bitcoins from.

# Technical Details

### Overview

The code was written based on the [Ledger Live App documentations](https://developers.ledger.com/docs/live-app/start-here/). As you can see there are two sections in the documentation: [DApp](https://developers.ledger.com/docs/dapp/process/) and [Non-DApp](https://developers.ledger.com/docs/non-dapp/introduction/) - both describe two different ways of embedding an application into the Ledger Live Discover section. A first natural choice in our case would be the `DApp` section, since our website is a Dapp. Unfortunately, that is not the case, because from my experience and research it looks like it was not possible to do a bitcoin transaction there. This is why we choose the second option, which allows to use [Wallet-API](https://wallet.api.live.ledger.com/). With the help of this API we are able to do bitcoin and eth transactions, and also interact with eth contracts.

The Wallet-API also has two sections in the docs: [Core-API](https://wallet.api.live.ledger.com/core) and [React-API](https://wallet.api.live.ledger.com/react), that uses Core-API under the hood. In our case we actually use both: React-API for connecting the eth/btc accounts and sending bitcoin transactions from one account to another (in our case to deposit address) and Core-Api to interact with eth contracts. Why?

The answer is that using only React-API would require us to reorganize [tBTC v2 SDK](https://github.com/keep-network/tbtc-v2/tree/main/typescript) just for the Ledger Live App functionality. The API for reacts needs raw data format of the ethereum transaction when we interact with the contract, and that can be obtained using [populateTransaction method](https://docs.ethers.org/v5/api/contract/contract/#contract-populateTransaction) from `ethers` lib, but we are not returning it in such form in our SDK. This is why we've decided to create a separate signer for this purpose - to avoid doing any changes in the SDK just for that feature and to not unnecessarily extend SDK responsibility.

### Ledger Live Ethereum Signer (wallet-api-core)

TBTC v2 SDK allows us to pass signer when initiating it. The signer must extend the `Signer` class from `ethers` lib and this is exactly what our Ledger Live Ethereum Signer do. It uses `wallet-api-core` lib under the hood. The signer [was placed in tbtc-v2 repo](https://github.com/keep-network/tbtc-v2/blob/releases/mainnet/typescript/v2.3.0/typescript/src/lib/utils/ledger.ts)

You can see a more detailed description of that signer, its purpose and explanation of how it works in https://github.com/keep-network/tbtc-v2/pull/743.

In our dApp we are requesting an eth account using `wallet-api-core-react` (see the subsection below) and then pass the account to the signer using [`setAccount` method](https://github.com/keep-network/tbtc-v2/blob/releases/mainnet/typescript/v2.3.0/typescript/src/lib/utils/ledger.ts#L65-L67).

### Connecting wallets and doing simple transactions (wallet-api-core-react)

The Ledger Live Ethereum Signer is used to integrate with eth contracts, but what about connecting the account to our dApp and sending some tokens from one account to another? This is where we use `wallet-api-core-react` and it's hooks.

In our dApp we have three custom hooks that use hooks from `wallet-api-core-react` under the hood:
- `useRequestBitcoinAccount`,
- `useRequestEthereumAccount`,
- `useSendBitcoinTransaction`.

The first two are pretty similar to the original ones (from the lib), but I've had to write a wrapper to it so that I can connect and disconnect `walletApiReactTransport` there. This is needed because our  Ledger Live Ethereum Signer uses different instance of the transport there, so if we won't disconnect one or another, a `no ongoing request` error might occur. Based on [the dosc](https://wallet.api.live.ledger.com/core/configuration#initializing-the-wallet-api-client) the transport should be disconnected when we are done to ensure the communication is properly closed.

The third one, `useSendBitcoinTransaction`,  is used to create a bitcoin transaction in a proper format that is required by `wallet-api-core-react`. The format for our bitcoin transaction looks like this:
```
const bitcoinTransaction = {
  family: "bitcoin", 
  amount: new BigNumber(100000000000000),
  recipient: "<bitcoin_address>",
};
```

Fields:
- `family` (string): The cryptocurrency family to which the transaction belongs. This could be 'ethereum', 'bitcoin', etc.
- `amount` (BigNumber): The amount of cryptocurrency to be sent in the transaction, represented in the smallest unit of the currency. For instance, in Bitcoin, an amount of 1 represents 0.00000001 BTC.
- `recipient` (string): The address of the recipient of the transaction.
- `nonce` (number, optional): This is the number of transactions sent from the sender's address.
- `data` (Buffer, optional): Input data of the transaction. This is often used for contract interactions.
- `gasPrice` (BigNumber, optional): The price per gas in wei.
- `gasLimit` (BigNumber, optional): The maximum amount of gas provided for the transaction.
- `maxPriorityFeePerGas `(BigNumber, optional): Maximum fee per gas to be paid for a transaction to be included in a block.
- `maxFeePerGas` (BigNumber, optional): Maximum fee per gas willing to be paid for a transaction.

_Source: https://wallet.api.live.ledger.com/appendix/transaction_

In our case, for our bitcoin transaction, we only need `family`, `amount` and `recipient`. We only use that to send bitcoins to deposit address, so we will use the deposit address as a `recipient` here.

Finally, to execute the transaction, we just pass the transaction object and id of the connected bitcoin account to [`useSignAndBroadcastTransaction` hook](https://wallet.api.live.ledger.com/react/hooks/useSignAndBroadcastTransaction).

### LedgerLiveAppContext

Connecting account in Ledger Live App is quite different than our actual one in the website. Normally, we use `web3react` for that, but in this case we need to use [`useRequestAccount` hook](https://wallet.api.live.ledger.com/react/hooks/useRequestAccount) form `wallet-api-client-react`. Because of that we need to store those accounts somewhere in our dApp, so I decided to create a `LedgerLiveAppContext` for that.

The context contain 5 properties:
```
interface LedgerLiveAppContextState {
  ethAccount: Account | undefined
  btcAccount: Account | undefined
  setEthAccount: (ethAccount: Account | undefined) => void
  setBtcAccount: (btcAccount: Account | undefined) => void
  ledgerLiveAppEthereumSigner: LedgerLiveEthereumSigner | undefined
}
```

As you can see we have `ethAccount` and `btcAccount` to store the connected accounts there. We can also set those account using `setEthAccount` and `setBtcAccount` methods, after we request it using our hook. The `ledgerLiveAppEthereumSigner` is an additional property that contains our signer for Ledger Live App. This way we will be able to set the account also in the signer.

### `useIsEmbed` hook

Like I said earlier, we use `isEmbed` query parameter to determine if the dApp is used in Ledger Live or not. I've created an `useIsEmbed` hook that saves that query parameter to local storage and the use it to detect if we should use all the functionalities for Ledger Live App or not.

### `useDetectEmbed` hook

Additional to above, I've created `useDetectEmbed` hook, that detects if application is Embed based on the `?embed=true` query parameter. Besides that, it also adds `?embed=true` query parameter to every url in the page. 

This is needed because with only `useIsEmbed` hook, the `?embed=true` flag was needed only in home page (`https://dashboard.threshold.network?embed=true`) which then was saved to local storage and the url was changed to `https://dashboard.threshold.network/tBTC/mint`. This worked well for Ledger Live, because we still used embed state when we refreshed the page, but was problematic for website, because adding `?embed=true` parameter to the url would broke the page and required the user to manually remove the `isEmbed` from local storage.

With `useDetectEmbed` hook every url will have `?embed=true` query parameter, if the `isEmbed` flag in local storage is set to true, meaning that adding the query param once to the url will be enough to keep it in every url. After refreshing the page it will be checked if the parameter is in the url, and then we will set the correct value in local storage. This means, that if user use `?embed=true` flag in the website, he would have to just remove that query string from the url to fix it. On the Ledger Live it will still persist the embed mode once the user refreshes the page.

### `useIsActive` hook

This is also a new hook here. His main purpose is to determine if, and what account is active. Up to this point we've used `useWeb3React` hook for that purpose, but in this case it won't work. So, under the hook, the `useIsActive` returns similar values to `useWeb3React` hook if the app is not embed, but if it is, then we return proper values based on the `LedgerLiveAppContext`.

### How it works with `threshold-ts` lib

I've actually manage to not do any changes in our `threshold-ts` lib. The way it works now is that when the `isEmbed` flag is set to true, we pass the Ledger Live Ethereum Signer as a `providerOrSigner` property.

This required me to change `getContract` and `getBlock` method though, so that they return the proper values when tthe `providerOrSigner` is and instance of `LedgerLiveEthereumSigner`.

# Read More

- [Ledger Live App documentation](https://developers.ledger.com/docs/live-app/start-here/)
- [Wallet-Api documentation](https://wallet.api.live.ledger.com/)

# How To Test

**Note: Right now it can't be tested on Ledger Live, because it does not support Sepolia. Please see last point in TODO list at the bottom of this description.**

Steps to Run it in as Ledger Live App:
1. Pull the newest changes from this branch
2. Run Ledger Live on your device
3. [Enable the developer mode](https://developers.ledger.com/docs/live-app/developer-mode/)
4. Go to Settings -> Developer
5. Go to `Add a local app` row and click `Browse`
6. Got to your project directory and choose [manifest-ledger-live-app.json](https://github.com/threshold-network/token-dashboard/blob/ledger-live-app/manifest-ledger-live-app.json) 
6. Click `Open`

# To Do List

Task list:
- [x] Display only the content of [tbtc subpage](https://dashboard.test.threshold.network/tBTC/mint) in our Live App with the use of `isEmbed` flag,
- [x] Write Ethereum Signer for Ledger Live App, which we can then pass to SDK v2,
- [x] Implement minting flow for Live App using SDK v2,
- [x] Implement unminting flow for Live App using SDK v2,
- [x] Implement resume deposit flow for Live App using SDK v2,
- [x] Allow user to input the BTC value that he wants to send to deposit address,
- [x] Move [Ledger Live App Signer](https://github.com/threshold-network/token-dashboard/blob/ledger-live-app/src/ledger-live-app-eth-signer/index.ts) to [tbtc-v2 repository](https://github.com/keep-network/tbtc-v2) - https://github.com/keep-network/tbtc-v2/pull/743,
  - [x] Import moved Ledger Live App Signer from `tbtc-v2.ts`
- [x] Add/Test wallet disconnect functionality (Found bug while doing that: https://github.com/threshold-network/token-dashboard/issues/669. It is not related to Ledger though.),
- [x] Fix small issues marked as TODO in the code,
- [x] Wait for https://github.com/threshold-network/token-dashboard/pull/654 to be merged an then:
  - [x] merge it into this branch and resovle conflits,
  - [x] test if everything works properly,
  - [x] fix any bugs that will appear;
- [x] Add extended description for this PR and make it ready for review (after https://github.com/threshold-network/token-dashboard/pull/654 is merged),
- [x] Fix review comments (if any),
- [x] Merge `main` into this branch to apply changes from https://github.com/threshold-network/token-dashboard/pull/605 (note: Unfortunately as of know Ledger Live App does not support Sepolia)
- [ ] Merge!

In the future:
- [ ] Write [Ledger Live App Plugin](https://developers.ledger.com/docs/dapp/requirements/) so we can display proper information on the Ledger device when revealing a deposit or requesting a redemption
- [ ] Implement/check if the plugin works on Sepolia. It's currently [under development](https://github.com/LedgerHQ/ledger-live/pull/5722).

kkosiorowska reviewed (changes requested):
I haven't tested it yet.  But I had a quick look and left a few minor comments.

I wonder if you have noticed, but some applications remember the connected account. I think it would be good to do that as well.

**Current behaviour**

https://github.com/threshold-network/token-dashboard/assets/23117945/4f131fc9-f73a-4a82-8c6f-ee503491c022

**Good behaviour**

https://github.com/threshold-network/token-dashboard/assets/23117945/b5974ef0-89cc-4005-94a8-b7819b1d13fb

r-czajkowski commented on manifest-ledger-live-app.json:
```diff
@@ -0,0 +1,40 @@
+  "name": "Threshold",
+  "url": "http://localhost:3000?embed=true",
+  "homepageUrl": "http://localhost:3000?embed=true",
+  "icon": "data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAD40lEQVR4AayXAWRVYRTH32sBAGCEspHqBQEFAAEFAAYmElK1giIAaEstULYRUU1tFSAEJDDYaxsqGBS59pRZe+/ddzo/3zv83btdt3T47zvvnPP9v/Pd79y78zVUzEzRHAwGI/J71DHhmHOsuC9z5K6DDBs+Ytw+KvPgaCr3nuKTdPEDop9wzDs6Vl+IXXC0gsf5lbNRtfPY9UGfNO3jQIh7Q/SxK7CJPwT7NFzBKQmlxcUQAejjjlUh6jpyIQ2Y2hBDiE1zQtqOMV0jEErsmPGUPO5dWUgXrRKNY4QD2XJ+uBtaW/zQMx8jUHZtur ...
```
Can we get the icon from url? Eg. `https://dashboard.threshold.network/favicon-T.png` to get the img from `public` dir.

michalsmiarowski replied:
1958cfa76029c02bef54196fb684d8c32e52d2dd

r-czajkowski commented on manifest-ledger-live-app.json:
```diff
@@ -0,0 +1,40 @@
+      "en": "tBTC is Threshold’s decentralized bridge to bring BTC to the Ethereum network; the only permissionless solution on the market today. Bridge your Bitcoin to Ethereum in a secure and trustless way to participate in DeFi."
+    }
+  },
+  "permissions": [
```
I can't find it in the docs only in the code: https://github.com/LedgerHQ/wallet-api/blob/0bd66f57e23e0f5704e2b780bd8e6eae7b1eba25/apps/wallet-api-tools/manifest.json#L23-L40

r-czajkowski commented on src/components/Navbar/index.tsx:
```diff
@@ -1,18 +1,40 @@
+    }
+  }
+
+  useEffect(() => {
+    if (ledgerLiveAccount) {
+      dispatch(walletConnected(ledgerLiveAccount.address))
+    }
+  }, [ledgerLiveAccount])
```
```suggestion
  const ledgerLiveAccountAddress = ledgerLiveAccount.address
  
  useEffect(() => {
    if (ledgerLiveAccountAddress) {
      dispatch(walletConnected(ledgerLiveAccountAddress))
    }
  }, [ledgerLiveAccountAddress, dispatch])
```

michalsmiarowski replied:
683115e2bdf9e6f186f6ba07ad79b9e6d35b0655

r-czajkowski commented on src/components/SubmitTxButton.tsx:
```diff
@@ -15,11 +17,21 @@ const SubmitTxButton: FC<Props> = ({
   const { isSdkInitializedWithSigner } = useIsTbtcSdkInitializing()
 
-  if (active && isSdkInitializedWithSigner) {
+  const connectWallet = () => {
+    if (isEmbed) {
+      requestAccount()
+    } else {
+      openModal(ModalType.SelectWallet)
+    }
+  }
```
We do the same in the `src/components/Navbar/index.ts`. I would consider creating a new hook like `useConnectWallet` or `useRequestAccount` to hanlde this logic.

michalsmiarowski replied:
84a6bf7613cf1bcd18ffd0c40fc2097dab54ec3d

michalsmiarowski commented on manifest-ledger-live-app.json:
```diff
@@ -0,0 +1,40 @@
+  "apiVersion": "2.0.0",
+  "manifestVersion": "1",
+  "branch": "main",
+  "categories": ["buy"],
```
a65ca816d757c2667fbd98fa1a2ddbfe7e1ca2d7

michalsmiarowski commented on manifest-ledger-live-app.json:
```diff
@@ -0,0 +1,40 @@
+  "platform": "all",
+  "apiVersion": "2.0.0",
+  "manifestVersion": "1",
+  "branch": "main",
```
a65ca816d757c2667fbd98fa1a2ddbfe7e1ca2d7

michalsmiarowski commented on src/App.tsx:
```diff
@@ -129,6 +132,12 @@ const useSubscribeToVendingMachineContractEvents = () => {
+  const params = new URLSearchParams(window.location.search)
+  if (params.get("embed")) {
+    enableIsEmbed()
+  }
```
03fc8c455ff217d8d9adfd4d31532cb8e4bde001 - was that what you mean?

michalsmiarowski replied:
I'm unresolving the conversation because I've acually changed it a little. See 6903e9030aa8bd9db36369c3a503f47a1bbe51d9

michalsmiarowski commented on src/components/TokenBalance.tsx:
```diff
@@ -95,8 +95,8 @@ const TokenBalance: FC<TokenBalanceProps & TextProps> = ({
-  const { active } = useWeb3React()
-  const shouldRenderTokenAmount = active
+  const { isActive } = useIsActive()
+  const shouldRenderTokenAmount = isActive
```
af42575cf15f249c43360df6587150ab34a3158f

michalsmiarowski commented on src/contexts/TransportProvider.tsx:
```diff
@@ -0,0 +1,53 @@
+  //     onMessage: undefined,
+  //     send: () => {},
+  //   }
+  // }
```
f65103a5d62ec581d15cee2df70f84358d7f82a5

michalsmiarowski commented:
> I haven't tested it yet. But I had a quick look and left a few minor comments.
> 
> I wonder if you have noticed, but some applications remember the connected account. I think it would be good to do that as well.
> 
> **Current behaviour**
> 
>  Screen.Recording.2023-11-29.at.12.07.37.mov 
> **Good behaviour**
> 
>  Screen.Recording.2023-11-29.at.12.18.20.mov

I believe this is how our dApp works right now on the website too. When you refresh the page you are not automatically connected to the wallet again. I don't remember why we've changed it, but I think there were some bugs that were happening and we implemented it as a workaround 🤔 

I think this could be a separate issue - to implement an automatic wallet connection with the previously used account once the page is refreshed @lukasz-zimnoch

michalsmiarowski commented on manifest-ledger-live-app.json:
```diff
@@ -0,0 +1,40 @@
+  "id": "threshold",
+  "name": "Threshold",
+  "url": "http://localhost:3000?embed=true",
+  "homepageUrl": "http://localhost:3000?embed=true",
```
Changing the URL won't fire the embed flg unfortunately 😞 Can I ask where that screenshot comes from though? 😅  I didn't find any occurences of the website link in our tbtc section of the dApp.

kkosiorowska replied:
Oh sorry I missed that message. You have to click the below icon.
<img width="1444" alt="Screenshot 2023-12-11 at 15 02 23" src="https://github.com/threshold-network/token-dashboard/assets/23117945/b7650d71-67ef-42bb-b5fa-a99be511a05a">

michalsmiarowski replied:
Thanks! Great catch! Fixed it here: 790c240c59bd3ba3422e6423381083d17862e070

kkosiorowska commented on src/contexts/TokenContext.tsx:
```diff
@@ -39,7 +40,8 @@ export const TokenContextProvider: React.FC = ({ children }) => {
   const nuConversion = useVendingMachineRatio(Token.Nu)
   const keepConversion = useVendingMachineRatio(Token.Keep)
-  const { active, chainId, account } = useWeb3React()
+  const { chainId } = useWeb3React()
```
Non-blocking comment
`chainId` can be also take from `useIsActive`

michalsmiarowski replied:
2e5ce7d943599a9662811d28ee7b93f68df3ebb0

kkosiorowska reviewed (commented):
Overall it looks good to me. ✨Only small non-blocking comments.

However, it would be good if @r-czajkowski also took a look at this solution to make sure that this PR doesn't break anything. He knows the app better.

michalsmiarowski commented on src/components/Navbar/index.tsx:
```diff
@@ -1,18 +1,34 @@
+    if (ledgerLiveAccountAddress) {
+      dispatch(walletConnected(ledgerLiveAccountAddress || ""))
+    }
+  }, [ledgerLiveAccountAddress])
```
aa77615b5854065f868189231ccb5630e4ef2415

michalsmiarowski commented on src/components/Navbar/index.tsx:
```diff
@@ -1,18 +1,34 @@
+
+  useEffect(() => {
+    if (ledgerLiveAccountAddress) {
+      dispatch(walletConnected(ledgerLiveAccountAddress || ""))
```
I think you're right. Thanks! Fixed in aa77615b5854065f868189231ccb5630e4ef2415

michalsmiarowski commented on src/components/SubmitTxButton.tsx:
```diff
@@ -15,11 +14,15 @@ const SubmitTxButton: FC<Props> = ({
   const { isSdkInitializedWithSigner } = useIsTbtcSdkInitializing()
+  const connectWallet = useConnectWallet()
 
-  if (active && isSdkInitializedWithSigner) {
```
I've actually didn't remove it. It's [right there](https://github.com/threshold-network/token-dashboard/pull/655/files/84a6bf7613cf1bcd18ffd0c40fc2097dab54ec3d#diff-9bffa157d0de0d9b2b6c0d255b374afe06eef491d3dee3b07f773032a9adb308R25) 😇 It looks like this in the diff because I've also added `onConnectWalletClick` funcion above it.
